server/vulnerabilities/nvd/sync/cve_syncer_test.go

```diff
@@ -0,0 +1,145 @@
+	"github.com/go-kit/log"
+	"github.com/google/go-cmp/cmp"
+	"github.com/pandatix/nvdapi/v2"
+	"github.com/stretchr/testify/require"
+)
+
+var (
+	legacyCVEFeedsDir = filepath.Join("testdata", "cve", "legacy_feeds")
+	api20CVEDir       = filepath.Join("testdata", "cve", "api_2.0")
+)
+
+func TestStoreCVEsLegacyFormat(t *testing.T) {
+	year := 2023
+	t.Run(fmt.Sprintf("%d", year), func(t *testing.T) {
+		// Load CVEs from legacy feed.
+		legacyCVEFilePath := filepath.Join(legacyCVEFeedsDir, fmt.Sprintf("%d.json.gz", year))
+		var legacyCVEs schema.NVDCVEFeedJSON10
+		loadJSONGz(t, legacyCVEFilePath, &legacyCVEs)
+
+		// Load CVEs from new API 2.0 feed.
+		api20CVEFilePath := filepath.Join(api20CVEDir, fmt.Sprintf("%d.json.gz", year))
+		var api20CVEs []nvdapi.CVEItem
+		loadJSONGz(t, api20CVEFilePath, &api20CVEs)
+
+		// Setup map of legacy CVEs.
+		legacyVulns := make(map[string]schema.NVDCVEFeedJSON10DefCVEItem) // key is the CVE ID.
+		for _, legacyVuln := range legacyCVEs.CVEItems {
+			legacyVulns[legacyVuln.CVE.CVEDataMeta.ID] = *legacyVuln
+		}
+
+		// Convert CVEs fetched using the new API 2.0 to the old legacy feeds format
+		// and compare them with the corresponding fetched legacy CVE.
+		var (
+			vulnsNotFoundInLegacy []string
+			mismatched            []string
+			matched               = 0
+		)
+		for _, api20Vuln := range api20CVEs {
+			convertedLegacyVuln := convertAPI20CVEToLegacy(api20Vuln, log.NewNopLogger())
+			legacyVuln, ok := legacyVulns[*api20Vuln.CVE.ID]
+			if !ok {
+				vulnsNotFoundInLegacy = append(vulnsNotFoundInLegacy, *api20Vuln.CVE.ID)
+				continue
+			}
+
+			if compareVulnerabilities(legacyVuln, *convertedLegacyVuln) {
+				matched++
+			} else {
+				mismatched = append(mismatched, *api20Vuln.CVE.ID)
+			}
+		}
+		matchRate := float64(matched) / float64(len(api20CVEs))
+		require.Greater(t, matchRate, .99)
+		t.Logf("%d: CVEs count: %d, match count: %d, match rate: %f", year, len(api20CVEs), matched, matchRate)
+		// TODO(lucas): Review these CVEs to check they are a-ok to be skipped.
+		t.Logf("%d: Vulnerabilities not found in legacy store: %s", year, strings.Join(vulnsNotFoundInLegacy, ", "))
+		t.Logf("%d: Vulnerabilities that mismatch from legacy store: %s", year, strings.Join(mismatched, ", "))
+	})
+}
+
+func compareVulnerabilities(v1 schema.NVDCVEFeedJSON10DefCVEItem, v2 schema.NVDCVEFeedJSON10DefCVEItem) bool {
+	clearDifferingFields := func(v *schema.NVDCVEFeedJSON10DefCVEItem) {
+		sort.Slice(v.CVE.References.ReferenceData, func(i, j int) bool {
+			return v.CVE.References.ReferenceData[i].URL < v.CVE.References.ReferenceData[j].URL
+		})
+		sortChildren(v.Configurations.Nodes)
+		for _, referenceData := range v.CVE.References.ReferenceData {
+			referenceData.Refsource = ""
+			referenceData.Name = referenceData.URL
+		}
+		// These fields do not match because we are comparing different sources (old feeds and new API).
+		v.LastModifiedDate = ""
+		v.PublishedDate = ""
```

Comment:
Do we show these dates on the frontend? If so, shouldn't the dates be close to the new API dates?

Reply:
Good catch. It seems that even if I remove this, the match rate is > 0.99 (it only differs on a few CVEs from the testdata).